lib/chewy/strategy/delayed_sidekiq.rb

```diff
@@ -0,0 +1,152 @@
+      FALLBACK_FIELDS = 'all'
+      FIELDS_IDS_SEPARATOR = ';'
+      IDS_SEPARATOR = ','
+
+      class Worker
+        include ::Sidekiq::Worker
+
+        def perform(type, score, options = {})
+          options[:refresh] = !Chewy.disable_refresh_async if Chewy.disable_refresh_async
+
+          ::Sidekiq.redis do |redis|
+            timechunks_key = "#{KEY_PREFIX}:#{type}:timechunks"
+            timechunk_keys = redis.zrangebyscore(timechunks_key, -1, score)
+            members = timechunk_keys.flat_map { |timechunk_key| redis.smembers(timechunk_key) }.compact
+
+            # extract ids and fields & do the reset of records
+            ids, fields = extract_ids_and_fields(members)
+            options[:update_fields] = fields if fields
+
+            index = type.constantize
+            index.strategy_config.delayed_sidekiq.reindex_wrapper.call do
+              options.any? ? index.import!(ids, **options) : index.import!(ids)
+            end
+
+            redis.del(timechunk_keys)
+            redis.zremrangebyscore(timechunks_key, -1, score)
+          end
+        end
+
+      private
+
+        def extract_ids_and_fields(members)
+          ids = []
+          fields = []
+
+          members.each do |member|
+            member_ids, member_fields = member.split(FIELDS_IDS_SEPARATOR).map { |v| v.split(IDS_SEPARATOR) }
+            ids |= member_ids
+            fields |= member_fields
+          end
+
+          fields = nil if fields.include?(FALLBACK_FIELDS)
+
+          [ids.map(&:to_i), fields]
+        end
+      end
+
+      class Scheduler
+        def initialize(type, ids, options = {})
+          @type = type
+          @ids = ids
+          @options = options
+        end
+
+        def postpone
+          ::Sidekiq.redis do |redis|
+            # warning: Redis#sadd will always return an Integer in Redis 5.0.0. Use Redis#sadd? instead
+            if redis.respond_to?(:sadd?)
+              redis.sadd?(timechunk_key, serialize_data)
+            else
+              redis.sadd(timechunk_key, serialize_data)
+            end
+            redis.expire(timechunk_key, expire_in)
+
+            unless redis.zrank(timechunks_key, timechunk_key)
+              redis.zadd(timechunks_key, at, timechunk_key)
+              redis.expire(timechunks_key, expire_in)
+
+              ::Sidekiq::Client.push(
+                'queue' => sidekiq_queue,
+                'at' => at + margin,
+                'class' => Chewy::Strategy::DelayedSidekiq::Worker,
+                'args' => [type_name, at]
+              )
+            end
+          end
+        end
+
+      private
+
+        attr_reader :type, :ids, :options
+
+        def expire_in
+          latency * 10 # avoid redis growing in case of dead worker
+        end
+
+        def at
+          @at ||= begin
+            schedule_at = latency.seconds.from_now.to_f
+
+            (schedule_at - (schedule_at % latency)).to_i
+          end
```

Comment:
Yep. I don't want to schedule tons of jobs. I want to have a single job during the specific time window based on the `latency` value.
Let's take a look at this example:
```ruby
def simulate(latency) # `at` code wrapped by dummy `simulate` method.
     schedule_at = latency.seconds.from_now.to_f
     (schedule_at - (schedule_at % latency)).to_i
end

latency = 2 # in seconds

# simulate sequential execution of the method
10.times { |i| puts "current_time: #{Time.current.to_i}, at: #{simulate(latency)}"; sleep(0.7) }                                                                                                  
current_time: 1679347458, at: 1679347460
current_time: 1679347459, at: 1679347460
current_time: 1679347459, at: 1679347460
current_time: 1679347460, at: 1679347462
current_time: 1679347461, at: 1679347462
current_time: 1679347461, at: 1679347462
current_time: 1679347462, at: 1679347464
current_time: 1679347463, at: 1679347464
current_time: 1679347463, at: 1679347464
current_time: 1679347464, at: 1679347466
```
now imagine that `reindex` is triggered at `current_time`. I do not schedule reindex at current_time, instead of that I accumulate all `ids` to reindex during the latency time window and schedule the reindex job at `at` value.
This is a very simple trick that is a core of the `delayed_sidekiq` strategy.

Reply:
makes sense, missed, thanks for clarification